### PR TITLE
chore: Migrate API_KEY_MATCHER and log_schema() into DatadogAgentSource

### PR DESCRIPTION
**Pull request by blt**
This commit removes a lazy_static and call to a lazy_static from the
DatadogAgentSource. As a part of #8512 I spent some time trying to remove warp
from the agent source to see its effect and found that, on the whole, it would
not really be worth all that much. warp might bloat our stack traces somewhat
but is lean internally. That said, I did find that I could remove some
serialization from the source, primarily by migrating globally retrieved types
into the source struct itself.

See also #8472

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

File: src/sources/datadog/agent.rs [hunk 1]
```diff
@@ -10,19 +10,12 @@ use crate::{
...
-use lazy_static::lazy_static;
 use regex::Regex;
 use serde::{Deserialize, Serialize};
 use std::{collections::HashMap, net::SocketAddr, sync::Arc};
 use vector_core::event::LogEvent;
-use warp::http::StatusCode;
-
 use warp::http::HeaderMap;
-
-lazy_static! {
```
Comment (spencergilbert) [inline src/sources/datadog/agent.rs:22]: Removing `lazy_static`s seems to be a common theme with your perf PRs - what's the issue with them?

**Review by spencergilbert — APPROVED** [review]

File: src/sources/datadog/agent.rs (hunk 1 above)
Comment (blt) [reply src/sources/datadog/agent.rs:22]: So the problem with `lazy_static` for vector is that what's actually static here -- in the sense of appearing in the static portion of the executable -- is a mutable place in memory for putting a value into, potentially with a race between multiple threads. Writing and read, then, implies an atomic barrier has to be passed through which on x86 will cause threads to queue. 

Used to be lazy_static was the only game in town. Now we should prefer the use of once_cell where we _do_ actually need a globalish variable or, better, just put the data local to whatever struct needs to make use of it.

File: src/sources/datadog/agent.rs [hunk 2]
```diff
@@ -73,9 +64,71 @@ impl SourceConfig for DatadogAgentConfig {
...
+    api_key_matcher: Regex,
+    log_schema_timestamp_key: &'static str,
+    log_schema_source_type_key: &'static str,
+}
+
+impl DatadogAgentSource {
+    fn new(store_api_key: bool) -> Self {
+        Self {
+            store_api_key,
+            api_key_matcher: Regex::new(r"^/v1/input/(?P<api_key>[[:alnum:]]{32})/??").unwrap(),
```
Comment (jszwedko) [inline src/sources/datadog/agent.rs:79]:
```suggestion
            api_key_matcher: Regex::new(r"^/v1/input/(?P<api_key>[[:alnum:]]{32})/??").expect("static regex always compiles"),
```

I realize we were using `unwrap()` prior; I just like that you can provide a reason with `expect`.

**Review by jszwedko — APPROVED** [review]

File: src/sources/datadog/agent.rs (hunk 1 above)
Comment (spencergilbert) [reply src/sources/datadog/agent.rs:22]: Thanks!

**Pull request merged**